transports/websocket/src/lib.rs

```diff
@@ -118,78 +120,94 @@ where
 pub type InnerStream<T, E> = BoxStream<'static, Result<ListenerEvent<InnerFuture<T, E>>, Error<E>>>;
 
 /// Type alias corresponding to `framed::WsConfig::Dial` and `framed::WsConfig::ListenerUpgrade`.
-pub type InnerFuture<T, E> = BoxFuture<'static, Result<BytesConnection<T>, Error<E>>>;
+pub type InnerFuture<T, E> = BoxFuture<'static, Result<Connection<T>, Error<E>>>;
 
 /// Function type that wraps a websocket connection (see. `wrap_connection`).
-pub type WrapperFn<T> = fn(BytesConnection<T>, ConnectedPoint) -> RwStreamSink<BytesConnection<T>>;
+pub type WrapperFn<T> = fn(Connection<T>, ConnectedPoint) -> RwStreamSink<BytesConnection<T>>;
 
 /// Wrap a websocket connection producing data frames into a `RwStreamSink`
 /// implementing `AsyncRead` + `AsyncWrite`.
-fn wrap_connection<T>(c: BytesConnection<T>, _: ConnectedPoint) -> RwStreamSink<BytesConnection<T>>
+fn wrap_connection<T>(c: Connection<T>, _: ConnectedPoint) -> RwStreamSink<BytesConnection<T>>
 where
     T: AsyncRead + AsyncWrite + Send + Unpin + 'static
 {
-    RwStreamSink::new(c)
+    RwStreamSink::new(BytesConnection(c))
+}
+
+/// The websocket connection.
+#[derive(Debug)]
+pub struct BytesConnection<T>(Connection<T>);
+
+impl<T> Stream for BytesConnection<T>
+where
+    T: AsyncRead + AsyncWrite + Send + Unpin + 'static
+{
+    type Item = io::Result<BytesMut>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        loop {
+            match ready!(self.0.try_poll_next_unpin(cx)?) {
+                Some(framed::IncomingData::Binary(d)) => return Poll::Ready(Some(Ok(d))),
+                None => return Poll::Ready(None),
+                _ => {}
+            }
+        }
+    }
+}
+
+impl<T> Sink<BytesMut> for BytesConnection<T>
```

Comment:
As a basic comment (not requesting any change), we could also provide multiple implementations of the `Sink` but with different item types (text, ping, bytes).